File: Graph/Number Of Enclaves/Number_Of_Enclaves.cpp
// Problem Name: Number Of Enclaves
// Problem Link: https://www.geeksforgeeks.org/problems/number-of-enclaves/0

class Solution {
  public:
    int dx[4] = {-1,0,1,0};
    int dy[4] = {0,1,0,-1};
    void dfs(int i,int j,vector<vector<int>>& grid,vector<vector<int>>& vis){
        vis[i][j] = 1;
        for(int k = 0;k<4;k++){
            int row = i+dx[k];
            int col = j+dy[k];
            if(row>=0 && row<grid.size() && col>=0 && col<grid[0].size() && !vis[row][col] && grid[row][col]==1){
                dfs(row,col,grid,vis);
            }
        }
    }
    
    int numberOfEnclaves(vector<vector<int>> &grid) {
        // Code here
        int n = grid.size();
        int m = grid[0].size();
        vector<vector<int>> vis(n,vector<int>(m,0));
        // First row
        for(int i = 0;i<m;i++){
            if(!vis[0][i] && grid[0][i]==1){
                dfs(0,i,grid,vis);
            }
        }
        
        // Last row
        for(int i = 0;i<m;i++){
            if(!vis[n-1][i] && grid[n-1][i]==1){
                dfs(n-1,i,grid,vis);
            }
        }
        
        // First col
        for(int i = 0;i<n;i++){
            if(!vis[i][0] && grid[i][0]==1){
                dfs(i,0,grid,vis);
            }
        }
        
        // Last col
        for(int i = 0;i<n;i++){
            if(!vis[i][m-1] && grid[i][m-1]==1){
                dfs(i,m-1,grid,vis);
            }
        }
        
        int ans = 0;
        for(int i = 0;i<n;i++){
            for(int j = 0;j<m;j++){
                if(!vis[i][j] && grid[i][j]==1){
                    ans++;
                }
            }
        }
        return ans;
    }
};

File: Graph/Number Of Enclaves/Number_Of_Enclaves.java
// Problem Name: Number Of Enclaves
// Problem Link: https://www.geeksforgeeks.org/problems/number-of-enclaves/0 / https://leetcode.com/problems/number-of-enclaves/

class Solution {
    int[] dx = { -1, 0, 1, 0 };
    int[] dy = { 0, 1, 0, -1 };

    public void dfs(int i, int j, int[][] grid, int[][] vis) {
        vis[i][j] = 1;
        for (int k = 0; k < 4; k++) {
            int row = i + dx[k];
            int col = j + dy[k];
            if (row >= 0 && row < grid.length && col >= 0 && col < grid[0].length && vis[row][col] == 0
                    && grid[row][col] == 1) {
                dfs(row, col, grid, vis);
            }
        }
    }

    public int numEnclaves(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        int[][] vis = new int[m][n];
        // First Column
        for (int i = 0; i < m; i++) {
            if (vis[i][0] == 0 && grid[i][0] == 1) {
                dfs(i, 0, grid, vis);
            }
        }
        // Last Column
        for (int i = 0; i < m; i++) {
            if (vis[i][n - 1] == 0 && grid[i][n - 1] == 1) {
                dfs(i, n - 1, grid, vis);
            }
        }
        // First Row
        for (int j = 0; j < n; j++) {
            if (vis[0][j] == 0 && grid[0][j] == 1) {
                dfs(0, j, grid, vis);
            }
        }
        // Last Row
        for (int j = 0; j < n; j++) {
            if (vis[m - 1][j] == 0 && grid[m - 1][j] == 1) {
                dfs(m - 1, j, grid, vis);
            }
        }

        int ans = 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (vis[i][j] == 0 && grid[i][j] == 1) {
                    ans++;
                }
            }
        }
        return ans;
    }
}